packages/main/src/plugin/provider-registry.ts

```diff
@@ -1264,4 +1267,45 @@ export class ProviderRegistry {
       this.apiSender.send('provider-change', {});
     });
   }
+
+  async shellInProviderConnection(
+    internalProviderId: string,
+    providerConnectionInfo: ProviderContainerConnectionInfo | ProviderKubernetesConnectionInfo,
+    onData: (data: string) => void,
+    onError: (error: string) => void,
+    onEnd: () => void,
+  ): Promise<{ write: (param: string) => void; resize: (dimensions: ProviderConnectionShellDimensions) => void }> {
+    try {
+      const containerConnection = this.getMatchingConnectionFromProvider(internalProviderId, providerConnectionInfo);
+      let shellAccess: ProviderConnectionShellAccess | undefined;
+      let connection: ProviderConnectionShellAccessSession | undefined;
+      if (this.isContainerConnection(containerConnection) && providerConnectionInfo.status === 'started') {
+        shellAccess = containerConnection.shellAccess;
+        connection = shellAccess?.open();
+        connection?.onData(data => {
+          onData(data.data);
+        });
+        connection?.onError(error => {
+          onError(error.error);
+        });
+        connection?.onEnd(onEnd);
+      }
+
+      return {
+        write: (data: string): void => {
+          if (connection) {
+            connection.write(data);
+          }
```

Comment:
before (lines 1284-1291) you used `?`

```suggestion
          connection?.write(data);
```